view/frontend/web/js/track-checkout.js

```diff
@@ -0,0 +1,68 @@
+/* Track checkout funnel transition in magento's default checkout */
+
+var currentStep = '';
+
+function waitFor(condition, doWhenReady) {
+    var waitForInterval = setInterval(function () {
+        if (condition()) {
+            doWhenReady();
+            clearInterval(waitForInterval);
+        }
+    }, 50);
+}
+
+function trackFunnel(step) {
+    if (step === currentStep) {
+        // prevent duplicate tracking
+        return;
+    }
+    currentStep = step;
+    BoltTrack.recordEvent(step);
+}
+
+function setupListnerForShippingForm() {
+    var requiredFields = jQuery("#shipping-new-address-form .field._required");
+    var inputElements = [];
+    requiredFields.each(function(_, e) {
+        $elm = jQuery(e).find("input,select");
+        inputElements.push($elm);
+        $elm.change(function() {
+           var complete = true;
+           for (var i = 0; i < inputElements.length; i++) {
+               if (!inputElements[i].val() && inputElements[i].attr("name") !== "region") {
+                   complete = false;
+                   break;
+               }
+           }
+           if (complete) {
+               trackFunnel("onShippingAddressComplete")
+           }
+        });
+    });
+}
+
+function init() {
+    trackFunnel("onCheckoutStart");
+
+    waitFor(
+        function() { return jQuery("#shipping-new-address-form .field._required input,select").length === 9; },
+        setupListnerForShippingForm
+    );
+
+    window.addEventListener("hashchange", function() {
+        if (location.hash === "#payment") {
+            trackFunnel("onShippingOptionsComplete");
+        }
+    });
+
+    // on method can detect checkout button added after this query is executed.
+    jQuery(".page-wrapper").on("click", ".action.primary.checkout", function() {
+        trackFunnel("onPaymentSubmit")
+        return false;
+    });
+}
+
+waitFor(
```

Comment:
It looks like you're using plain JS to wait when JQuery is available and then using it for initializing the events. That might not work if JQuery isn’t available on the page as Magento uses RequireJS to inject the dependency (E.g. JQuery) and sometimes the global JQuery object isn’t available or they might use the $ variable as the global object for JQuery.

It might be better to create an ADM RequireJS module and inject the JQuery library as a dependency. Then you can remove the logic for waiting for JQuery to be available.

Reply:
cool I updated code to use requriejs